linear_regression_learner.go

```diff
@@ -0,0 +1,161 @@
+	instances	*base.DenseInstances
+	predictions	*base.DenseInstances
+}
+
+func (ll *LRLearner) Initialize() {
+	ll.minTrainValues = 32
+
+	// Initialize the DenseInstance
+	// Two attributes:
+	//	- CommandWindowSz: Independent variable
+	//	- Throughput: Dependent variable (ClassAttribute)
+	ll.attrs = make([]base.Attribute, 2)
+	ll.attrs[0] = base.NewFloatAttribute("CommandWindowSz")
+	ll.attrs[1] = base.NewFloatAttribute("Throughput")
+	ll.specs = make([]base.AttributeSpec, len(ll.attrs))
+	ll.instances	= base.NewDenseInstances()
+	for i, a := range ll.attrs {
+		spec := ll.instances.AddAttribute(a)
+		ll.specs[i] = spec
+	}
+	ll.instances.Extend(maxRows)
+	ll.instances.AddClassAttribute(ll.attrs[1])
+
+	// Make a copy in ll.predictions
+	ll.predictions = base.NewDenseCopy(ll.instances)
+	attrs := ll.predictions.AllAttributes()
+	ll.predictions.AddClassAttribute(attrs[1])
+
+	// Initialize the data set to -1 to distinguish updated training data with
+	// uninitialized data.
+	ll.data = make([]int64, maxRows)
+	for i := 0; i < maxRows; i++ {
+		ll.data[i] = -1
+	}
+}
+
+// Train() API just stores the data in the data array to be processed by the
+// regression api later in Predict() API
+func (ll *LRLearner) Train(cmdWinSize int32, tput int64) {
+	if ll.data[cmdWinSize] < 0 {
+		ll.inputSz++
+	}
+	ll.data[cmdWinSize] = tput
+}
+
+
+// This internal method populates the DenseInstances(golearn data structure) from
+// the training dataset. There are two DenseInstances:
+//	- ll.instances : Training dataset for the Regression module. Attributes initialized
+//	                 only for those rows which has training data
+//	- ll.predictions : The output prediction dataset. Attribute is initialized for all
+//	                   rows, so that we get predictions for all rows.
+//
+func (ll *LRLearner) populateTrainingData() {
+	instance_specs := base.ResolveAllAttributes(ll.instances)
+	prediction_specs := base.ResolveAllAttributes(ll.predictions)
+	row := 0
+	for i := 1; i < maxRows; i++ {
+		tput := ll.data[i]
+		ll.predictions.Set(prediction_specs[0], row, base.PackFloatToBytes(float64(i)))
+		if tput == -1 { continue }
+
+		ll.predictions.Set(prediction_specs[1], row, base.PackFloatToBytes(float64(tput)))
+
+		ll.instances.Set(instance_specs[0], row, base.PackFloatToBytes(float64(i)))
+		ll.instances.Set(instance_specs[1], row, base.PackFloatToBytes(float64(tput)))
+		row++
+	}
+}
+
+// Internal method which performs the linear regression.
+// Returns: DenseInstance predictions. This DenseInstance has only one attribute and that is
+//	    the ClassAtribute - Throughput.
+//
+func (ll *LRLearner) predict() (base.FixedDataGrid, error) {
+	if ll.inputSz < ll.minTrainValues {
+		return nil, NotEnoughDataError
+	}
+	ll.populateTrainingData()
+
+	lr := linear_models.NewLinearRegression()
+	err := lr.Fit(ll.instances)
+	if err != nil {
+		log(LogVVV, fmt.Sprintf("LinearRegression.Fit() failed: %v", err))
+		return nil, err
+	}
+
```

Comment:
Once fitting is done (via lr.Fit() above) , we have effectively a* x^2 + b * x + c polynomial expression, with a, b, and c computed by the golearn framework. And so there are two types of errors to look at: delta between the computed polynomial expression and the real throughput values for the training set, and the same delta but now for the testing set that must be kept separately from training. 

Both deltas must be fairly close to zero, for us to believe that we can safely use this  a* x^2 + b * x + c expression to calculate the maximum throughput.